### PR TITLE
Add Scaleway user_data module (Allows Cloud-init configuration on Scaleway instances)

### PR DESCRIPTION
**Pull request by remyleone**
##### SUMMARY

This merge request adds a scaleway module to push user_data to the Scaleway compute API. It is especially useful to perform a cloud-init configuration on the Scaleway compute instance through Ansible.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

- scaleway_user_data

##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (user_data dc2b7466df) last updated 2018/07/12 18:04:08 (GMT +200)
  config file = None
  configured module search path = [u'/Users/sieben/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/sieben/workspace/ansible/lib/ansible
  executable location = /Users/sieben/workspace/ansible/bin/ansible
  python version = 2.7.15 (default, Jun 17 2018, 12:46:58) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
```

### CODE REVIEW

**Comment by remyleone** [comment]
@mhalano This PR passes continuous integration and can be use to add cloud-init configuration tasks in your playbook. Let me know if you find it useful

**Comment by mhalano** [comment]
@sieben I'm having hard times making this module work. See my playbook:
```yaml
- hosts: localhost
  gather_facts: no
  tasks:

  - name: start the server
    scaleway_compute:
      name: test
      commercial_type: START1-XS
      region: par1
      image: 67375eb1-f14d-4f02-bb42-6119cecbde51
      organization: <my organization ID>
      state: running
    register: machine

  - name: add userdata
    scaleway_user_data:
      server_id: "{{ machine.msg.id }}"
      user_data:
        cloud-init: 'final_message: Hello World'
      region: par1
```

The machine is started (or keeping running) correctly, but the scaleway_user_data module always return this error:
```json
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "oauth_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "region": "par1", 
            "server_id": "4a5d3789-88c9-4222-bcd7-57b1753f90ee", 
            "timeout": 30, 
            "user_data": {
                "cloud-init": "final_message: Hello World"
            }
        }
    }, 
    "msg": "Error during user_data patching: 400 None"
}
```
Using curl do to the PATCH verb works well (minor problem with an exclamation mark).

**Comment by remyleone** [comment]
Your playbook works fine on my side.  Be sure to pull the last version from this branch

**Comment by sieben** [comment]
Your playbook works fine on my side.

**Comment by mhalano** [comment]
Now I'm using Ansible 2.7dev0 instead of Ansible 2.6.1 and worked fine. I still remains with a smail question about my image doesn't have cloud-init installed so user_data isn't useful now.

**Comment by remyleone** [comment]
@mhalano Can you use on the distribution listed here: https://www.scaleway.com/docs/how-to-use-cloud-init-to-configure-your-server-at-first-boot/ ? Let me know if I can help.

**Comment by remyleone** [comment]
@pilou- @maxamillion Could you tell me what you think?

File: lib/ansible/modules/cloud/scaleway/scaleway_user_data.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    description:
+    - Scaleway Compute instance ID of the server
+    required: true
+
+  oauth_token:
+    description:
+     - Scaleway OAuth token.
+    required: true
+
+  user_data:
```
Comment (maxamillion) [inline lib/ansible/modules/cloud/scaleway/scaleway_user_data.py]: The documentation seems to be lacking some details or clarification. I don't know what this parameter wants me to pass to it, what data type the arg should be, etc.

File: lib/ansible/modules/cloud/scaleway/scaleway_user_data.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+  timeout:
+    description:
+    - Timeout for API calls
+    required: false
+    default: 30
+'''
+
+EXAMPLES = '''
+- name: Create a server
```
Comment (abarbare) [inline lib/ansible/modules/cloud/scaleway/scaleway_user_data.py]: May be the name is not correct. This won't create a server.

File: lib/ansible/modules/cloud/scaleway/scaleway_user_data.py (hunk 1 above)
Comment (remyleone) [reply lib/ansible/modules/cloud/scaleway/scaleway_user_data.py]: I've added clarification. Tell me if you think it's clear enough :)

Comment (maxamillion) [reply lib/ansible/modules/cloud/scaleway/scaleway_user_data.py]: If the intent is for this to be a clout-init script, why is the param called `user_data`?

Comment (abarbare) [reply lib/ansible/modules/cloud/scaleway/scaleway_user_data.py]: `user_data` is a key value store for your Scaleway box. Here, `cloud-init` is one key but you can push any other information through it.

**Comment by abarbare** [comment]
I spotted another issue on this module for the `cloud-init` integration.
Currently, data pushed to `user_data` seems to be parsed in `yaml` and sent to python. Matter of fact, the data sent to `user_data` is not in plain text.

The example present in this module will create something like: 
```
~# scw-userdata cloud-init
"final_message: Hello World"
```
This kind of string is not cloud-init formatted, we need to have something like 
```
~# scw-userdata cloud-init
final_message: Hello World
```

Same for multiline text, we have something like that pushed in `user_data`
```
~# scw-userdata cloud-init
"#cloud-config
apt_update: true
apt_upgrade: true
"
```
I don't have yet a solution but I keep digging.

**Comment by abarbare** [comment]
I spotted another issue on this module for the `cloud-init` integration.
Currently, data pushed to `user_data` seems to be parsed in `yaml` and sent to python. Mather of fact, the data sent to `user_data` is not in plain text. 

The example present in this module will create something like: 
```
~# scw-userdata cloud-init
"final_message: Hello World"
```
This kind of string is not cloud-init formatted, we need to have something like 
```
~# scw-userdata cloud-init
final_message: Hello World
```

Same for multiline text, we have something like that pushed in `user_data`
```
~# scw-userdata cloud-init
"#cloud-config
apt_update: true
apt_upgrade: true
"
```
I don't have yet a solution but I keep digging.

File: lib/ansible/modules/cloud/scaleway/scaleway_user_data.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.basic import env_fallback
+from ansible.module_utils.scaleway import ScalewayAPI, SCALEWAY_LOCATION
+
+
+def patch_user_data(compute_api, server_id, key, value):
+    compute_api.module.debug("Starting patching user_data attributes")
+
+    path = "servers/%s/user_data/%s" % (server_id, key)
+    response = compute_api.patch(path=path, data=value, headers={"Content-Type": "text/plain"})
```
Comment (abarbare) [inline lib/ansible/modules/cloud/scaleway/scaleway_user_data.py]: To match the headers in https://github.com/sieben/ansible/blob/b119f2652a160707ee718566ad1e49a510935553/lib/ansible/module_utils/scaleway.py#L40
`Content-Type` must be `Content-type`

**Comment by remyleone** [comment]
@abarbare done for the Content-type

**Comment by remyleone** [comment]
@maxamillion @abarbare is there something else to do for this PR?

**Comment by maxamillion** [comment]
shipit

**Comment by remyleone** [comment]
@Spredzy I've rebased this module. Could you review it?

**Comment by maxamillion** [comment]
shipit

**Review by maxamillion — APPROVED** [review]

**Comment by remyleone** [comment]
shipit

**Comment by maxamillion** [comment]
rebuild_merge

**Comment by remyleone** [comment]
rebuild_merge

**Comment by remyleone** [comment]
@maxamillion I've fixed the CI. Could you merge?

**Comment by maxamillion** [comment]
rebuild_merge

**Pull request merged**